### PR TITLE
No pixel of a filled arc must ever be drawn multiple times

### PR DESCRIPTION
**Pull request by cmb69**
Otherwise we get artifacts regarding transparency. That happens with the
current implementation of gdImageFilledArc() unless gdChord or gdNoFill
are set. When gdPie is set, however, the filled arc is drawn in wedges,
which are polygons of three points, and so some overlap is natural.

To resolve the issue, we stick with the current algorithm of calculating the
wedges, but instead of drawing each polygon separately, we put the *relevant*
points in a large array, and draw a single polygon. That also is supposed to
improve the performance considerably.

Note that this modification will change the results when gdImageSetStyle()
or gdImageSetBrush() are used, but we believe that this modification is
also an improvement in this regard, even though it still might not make much
sense to use these functions with gdImageFilledArc().

### CODE REVIEW

File: src/gd.c [hunk 1]
```diff
@@ -1740,8 +1738,8 @@ BGD_DECLARE(void) gdImageArc (gdImagePtr im, int cx, int cy, int w, int h, int s
 BGD_DECLARE(void) gdImageFilledArc (gdImagePtr im, int cx, int cy, int w, int h, int s, int e,
 									int color, int style)
 {
-	gdPoint pts[3];
-	int i;
+	gdPoint pts[363];
```
Comment (cmb69) [inline src/gd.c]: This might be too demanding (~3KB on typical 32/64bit).

**Comment by cmb69** [comment]
I've run the PHP GD test suite with this patch, and at least 6 tests using imagefilledarc() produce exactly the same filled arc as before. These are all rather small arcs (at most the size of the one in the test case for this issue), but some quick tests with relatively large arcs (rx = ry = 1199) show reasonable results. Very large arcs are most likely rather cornered anyway, due to insufficient precision (only 360 points); we might consider to use floats/doubles in the future. Unfortunately, the libgd test suite doesn't have any tests regarding filled arcs yet.

Note that this patch is supposed to be unaffected by issue #258, as horizontal oscillation is filtered out by the modified gdImageFilledArc() implementation.

**Comment by cmb69** [comment]
For reference: the [PHP test case](https://github.com/php/php-src/blob/php-7.0.8/ext/gd/tests/imagecolorallocatealpha_basic.phpt) affected by this bug/fix produced before:
imagecolorallocatealpha_basic old

After applying the fix:
imagecolorallocatealpha_basic exp

**Comment by pierrejoye** [comment]
Can you make it for 2.2.3?

**Comment by cmb69** [comment]
> Can you make it for 2.2.3?

Yes. I'll also apply the fix to PHP's bundled libgd.

**Pull request merged**